File: android/app/src/main/kotlin/ai/gowtam/oak/services/ArtifactPinService.kt
package ai.gowtam.oak.services

import ai.gowtam.oak.features.artifact.ArtifactPinKind
import ai.gowtam.oak.networking.Endpoint
import ai.gowtam.oak.networking.OakApiClient
import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.wire.OakJson
import ai.gowtam.oak.wire.PinnedArtifactSummary
import android.util.Log
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.encodeToJsonElement

/**
 * Conversation pin strip (`/api/conversations/:id/artifact-pins`).
 * Signed-in only. Guest 401s fold to empty / error (AUTH-BR-1).
 */
interface ArtifactPinService {
    suspend fun list(conversationId: String): List<PinnedArtifactSummary>

    suspend fun create(
        conversationId: String,
        kind: ArtifactPinKind,
        title: String,
        snapshot: Any?,
    ): CreatePinResult

    suspend fun delete(conversationId: String, pinId: String): List<PinnedArtifactSummary>?
}

sealed interface CreatePinResult {
    data class Ok(val pin: PinnedArtifactSummary, val pinnedArtifacts: List<PinnedArtifactSummary>) : CreatePinResult
    data class Cap(val max: Int) : CreatePinResult
    data class Error(val message: String) : CreatePinResult
}

class LiveArtifactPinService(private val apiClient: OakApiClient) : ArtifactPinService {
    override suspend fun list(conversationId: String): List<PinnedArtifactSummary> {
        val endpoint = Endpoint(
            method = Endpoint.Method.GET,
            path = "/api/conversations/$conversationId/artifact-pins",
            requiresAuth = true,
        )
        return try {
            apiClient.send(endpoint, PinListEnvelope.serializer()).pins
        } catch (e: OakError) {
            Log.e(TAG, "pin list unavailable: ${e::class.simpleName}")
            emptyList()
        }
    }

    override suspend fun create(
        conversationId: String,
        kind: ArtifactPinKind,
        title: String,
        snapshot: Any?,
    ): CreatePinResult {
        val body = CreatePinBody(
            kind = kind.rawValue,
            title = title,
            snapshot = snapshotAsJson(snapshot),
        )
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/conversations/$conversationId/artifact-pins",
            body = Endpoint.jsonBody(CreatePinBody.serializer(), body),
            requiresAuth = true,
        )
        return try {
            val envelope = apiClient.send(endpoint, CreatePinEnvelope.serializer())
            CreatePinResult.Ok(envelope.pin, envelope.pinnedArtifacts)
        } catch (e: OakError.Http) {
            if (e.status == 409 || e.code == "pin_cap") {
                CreatePinResult.Cap(max = 5)
            } else {
                CreatePinResult.Error(e.message)
            }
        } catch (e: OakError) {
            Log.e(TAG, "pin create failed: ${e::class.simpleName}")
            CreatePinResult.Error("couldn't pin")
        }
    }

    override suspend fun delete(conversationId: String, pinId: String): List<PinnedArtifactSummary>? {
        val endpoint = Endpoint(
            method = Endpoint.Method.DELETE,
            path = "/api/conversations/$conversationId/artifact-pins/$pinId",
            requiresAuth = true,
        )
        return try {
            apiClient.send(endpoint, PinListEnvelope.serializer()).pinnedArtifacts
        } catch (e: OakError) {
            Log.e(TAG, "pin delete failed: ${e::class.simpleName}")
            null
        }
    }

    private fun snapshotAsJson(snapshot: Any?): JsonElement = when (snapshot) {
        null -> JsonNull
        is JsonElement -> snapshot
        is String -> OakJson.encodeToJsonElement(snapshot)
        else -> OakJson.encodeToJsonElement(snapshot.toString())
    }

    private companion object {
        const val TAG = "Oak.Pins"
    }
}

@Serializable
private data class CreatePinBody(
    val kind: String,
    val title: String,
    val snapshot: JsonElement,
)

@Serializable
private data class CreatePinEnvelope(
    val pin: PinnedArtifactSummary,
    val pinnedArtifacts: List<PinnedArtifactSummary> = emptyList(),
)

@Serializable
private data class PinListEnvelope(
    val pins: List<PinnedArtifactSummary> = emptyList(),
    val pinnedArtifacts: List<PinnedArtifactSummary> = emptyList(),
)

File: ios/OakApp/Services/ArtifactPinService.swift
import Foundation

enum ArtifactPinKind: String, Codable, Sendable, Equatable {
  case teamSheet = "team_sheet"
  case comparison
  case calc
}

struct PinnedArtifactSummary: Codable, Sendable, Equatable, Identifiable {
  let id: String
  let kind: ArtifactPinKind
  let title: String
  let createdAt: Int64

  enum CodingKeys: String, CodingKey {
    case id
    case kind
    case title
    case createdAt = "created_at"
  }
}

enum ArtifactPinError: Equatable, Sendable {
  case pinCap(max: Int)
  case unauthorized
  case failed
}

enum ArtifactPinCreateResult: Equatable, Sendable {
  case success(pin: PinnedArtifactSummary, pinnedArtifacts: [PinnedArtifactSummary])
  case failure(ArtifactPinError)
}

/// Conversation artifact pins (PIN-US-1–3). Never throws — guests / misses fold
/// to empty / failure so the strip can hide.
protocol ArtifactPinService: Sendable {
  func list(conversationId: String) async -> [PinnedArtifactSummary]
  func create(
    conversationId: String,
    kind: ArtifactPinKind,
    title: String,
    snapshot: Artifact
  ) async -> ArtifactPinCreateResult
  func get(conversationId: String, pinId: String) async -> Artifact?
  func delete(conversationId: String, pinId: String) async -> [PinnedArtifactSummary]
}

enum ArtifactPinEndpoints {
  static func list(conversationId: String) -> Endpoint {
    Endpoint(
      method: .get,
      path: "/api/conversations/\(conversationId)/artifact-pins",
      requiresAuth: true
    )
  }

  static func create(conversationId: String, body: CreatePinBody) -> Endpoint {
    Endpoint(
      method: .post,
      path: "/api/conversations/\(conversationId)/artifact-pins",
      body: body,
      requiresAuth: true
    )
  }

  static func get(conversationId: String, pinId: String) -> Endpoint {
    Endpoint(
      method: .get,
      path: "/api/conversations/\(conversationId)/artifact-pins/\(pinId)",
      requiresAuth: true
    )
  }

  static func delete(conversationId: String, pinId: String) -> Endpoint {
    Endpoint(
      method: .delete,
      path: "/api/conversations/\(conversationId)/artifact-pins/\(pinId)",
      requiresAuth: true
    )
  }
}

struct CreatePinBody: Encodable, Sendable {
  let kind: ArtifactPinKind
  let title: String
  let snapshot: PinSnapshotWire
}

struct PinSnapshotWire: Codable, Sendable, Equatable {
  let title: String
  let kind: String
  let payload: JSONValue
}

struct PinnedArtifactsListResponse: Decodable, Sendable {
  let pinnedArtifacts: [PinnedArtifactSummary]
}

struct CreatePinResponse: Decodable, Sendable {
  let pin: PinnedArtifactSummary
  let pinnedArtifacts: [PinnedArtifactSummary]
}

struct GetPinResponse: Decodable, Sendable {
  let pin: GetPinPayload
}

struct GetPinPayload: Decodable, Sendable {
  let id: String
  let kind: ArtifactPinKind
  let title: String
  let snapshot: JSONValue?
}

struct LiveArtifactPinService: ArtifactPinService {
  private let apiClient: OakAPIClient

  init(apiClient: OakAPIClient) {
    self.apiClient = apiClient
  }

  func list(conversationId: String) async -> [PinnedArtifactSummary] {
    do {
      let body = try await apiClient.send(
        ArtifactPinEndpoints.list(conversationId: conversationId),
        as: PinnedArtifactsListResponse.self
      )
      return body.pinnedArtifacts
    } catch {
      return []
    }
  }

  func create(
    conversationId: String,
    kind: ArtifactPinKind,
    title: String,
    snapshot: Artifact
  ) async -> ArtifactPinCreateResult {
    let body = CreatePinBody(
      kind: kind,
      title: title,
      snapshot: PinSnapshotWire(
        title: snapshot.title,
        kind: kind.rawValue,
        payload: .object([:])
      )
    )
    do {
      let response = try await apiClient.send(
        ArtifactPinEndpoints.create(conversationId: conversationId, body: body),
        as: CreatePinResponse.self
      )
      return .success(pin: response.pin, pinnedArtifacts: response.pinnedArtifacts)
    } catch OakError.http(let status, let code, _) where status == 409 || code == "pin_cap" {
      return .failure(.pinCap(max: 5))
    } catch OakError.unauthorized {
      return .failure(.unauthorized)
    } catch {
      return .failure(.failed)
    }
  }

  func get(conversationId: String, pinId: String) async -> Artifact? {
    do {
      let response = try await apiClient.send(
        ArtifactPinEndpoints.get(conversationId: conversationId, pinId: pinId),
        as: GetPinResponse.self
      )
      return Artifact(title: response.pin.title, content: .loading)
    } catch {
      return nil
    }
  }

  func delete(conversationId: String, pinId: String) async -> [PinnedArtifactSummary] {
    do {
      let body = try await apiClient.send(
        ArtifactPinEndpoints.delete(conversationId: conversationId, pinId: pinId),
        as: PinnedArtifactsListResponse.self
      )
      return body.pinnedArtifacts
    } catch {
      return []
    }
  }
}
